go/vt/sidecardb/sidecardb.go

```diff
@@ -0,0 +1,422 @@
+	case 0:
+		log.Infof("doesSidecarDBExist: not found")
+		return false, nil
+	case 1:
+		log.Infof("doesSidecarDBExist: found")
+		return true, nil
+	default:
+		log.Errorf("found too many rows for sidecarDB %s: %d", SidecarDBName, len(rs.Rows))
+		return false, fmt.Errorf("found too many rows for sidecarDB %s: %d", SidecarDBName, len(rs.Rows))
+	}
+}
+
+func (si *schemaInit) createSidecarDB() error {
+	_, err := si.exec(si.ctx, CreateSidecarDatabaseQuery, 1, false)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Infof("createSidecarDB: %s", CreateSidecarDatabaseQuery)
+	return nil
+}
+
+// sets db of current connection, returning the currently selected database
+func (si *schemaInit) setCurrentDatabase(dbName string) (string, error) {
+	rs, err := si.exec(si.ctx, SelectCurrentDatabaseQuery, 1, false)
+	if err != nil {
+		return "", err
+	}
+	if rs == nil || rs.Rows == nil { // we get this in tests
+		return "", nil
+	}
+	currentDB := rs.Rows[0][0].ToString()
+	if currentDB != "" { // while running tests we can get currentDB as empty
+		_, err = si.exec(si.ctx, fmt.Sprintf("use %s", dbName), 1, false)
+		if err != nil {
+			return "", err
+		}
+	}
+	return currentDB, nil
+}
+
+// gets existing schema of table
+func (si *schemaInit) getCurrentSchema(tableName string) (string, error) {
+	var currentTableSchema string
+
+	rs, err := si.exec(si.ctx, fmt.Sprintf(ShowCreateTableQuery, tableName), 1, false)
+	if err != nil {
+		log.Errorf("Error getting table schema for %s: %+v", tableName, err)
+		return "", err
+	}
+	if len(rs.Rows) > 0 {
+		currentTableSchema = rs.Rows[0][1].ToString()
+	}
+	return currentTableSchema, nil
+}
+
+// finds diff that needs to be applied to current table schema to get the desired one. Will be an empty string if they match.
+// could be a create statement if the table does not exist or an alter if table exists but has a different schema
+func (si *schemaInit) findTableSchemaDiff(current, desired string) (string, error) {
+	hints := &schemadiff.DiffHints{
+		TableCharsetCollateStrategy: schemadiff.TableCharsetCollateIgnoreAlways,
+	}
+	diff, err := schemadiff.DiffCreateTablesQueries(current, desired, hints)
+	if err != nil {
+		return "", err
+	}
+
+	var tableAlterSQL string
+	if diff != nil {
+		tableAlterSQL = diff.CanonicalStatementString()
+
+		// temp logging to debug any eventual issues around the new schema init, should be removed in v17
+		log.Infof("Current schema:\n%s", current)
+		if tableAlterSQL == "" {
+			log.Infof("No alter needed")
+		} else {
+			log.Infof("Alter query:\n%s", tableAlterSQL)
+		}
+	}
+
+	return tableAlterSQL, nil
+}
+
+// createOrUpgradeTable first checks if the table exist, in which case it runs the create script provided in
+// the schema directory. If the table exists, schemadiff is used to compare the existing schema with the desired one.
+// If it needs to be altered then we run the alter script.
+func (si *schemaInit) createOrUpgradeTable(table *sidecarTable) error {
```

Comment:
Changed the name as suggested. Also, as suggested above, refactored the code to always look for table existence, and rather than have separate code paths for create/alter, use schema diff to provide the ddl to apply.